File: populate_map.js
#!/usr/bin/env node
/**
 * Populate a SafeRoom3D map with aesthetically placed monsters and props.
 * Uses the map's tile data to intelligently place content.
 */

const fs = require('fs');
const zlib = require('zlib');
const path = require('path');

// Monster types by difficulty tier
const TIER1_MONSTERS = ["dungeon_rat", "slime", "goblin"];
const TIER2_MONSTERS = ["goblin_shaman", "goblin_thrower", "spider", "mushroom", "bat"];
const TIER3_MONSTERS = ["skeleton", "wolf", "lizard", "eye", "badlama"];
const TIER4_MONSTERS = ["crawler_killer", "shadow_stalker", "mimic", "flesh_golem"];
const TIER5_MONSTERS = ["living_armor", "plague_bearer", "lava_elemental", "void_spawn"];

const BOSSES = ["skeleton_lord", "dragon_king", "spider_queen", "the_butcher", "mordecai", "mongo"];

// Prop types by theme
const DUNGEON_PROPS = ["barrel", "crate", "pot", "torch", "bone_pile", "skull_pile", "rubble_heap"];
const TREASURE_PROPS = ["chest", "treasure_chest", "scattered_coins", "ancient_scroll"];
const SPOOKY_PROPS = ["blood_pool", "coiled_chains", "manacles", "discarded_sword", "forgotten_shield"];
const NATURE_PROPS = ["moss_patch", "glowing_mushrooms", "water_puddle", "thorny_vines"];
const CAMP_PROPS = ["campfire", "abandoned_campfire", "rat_nest", "moldy_bread"];

function decodeTileData(tileDataBase64, width, depth) {
    // Decode base64
    const compressed = Buffer.from(tileDataBase64, 'base64');
    // Decompress gzip
    const decompressed = zlib.gunzipSync(compressed);

    // Initialize tiles array
    const tiles = Array.from({ length: width }, () => Array(depth).fill(0));

    if (decompressed.length < 8) return tiles;

    // Read header (stored width and depth as little-endian int32)
    const storedWidth = decompressed.readInt32LE(0);
    const storedDepth = decompressed.readInt32LE(4);
    console.log(`Stored dimensions: ${storedWidth}x${storedDepth}`);

    // RLE decode starting after 8-byte header
    let dataIndex = 8;
    let x = 0, z = 0;

    while (dataIndex < decompressed.length && z < depth) {
        const firstByte = decompressed[dataIndex++];
        let value, count;

        if ((firstByte & 0x80) !== 0) {
            // Multi-byte run: marker byte (bit 7 = 1) + 2-byte count
            value = firstByte & 0x01;
            if (dataIndex + 1 >= decompressed.length) break;
            count = decompressed[dataIndex] | (decompressed[dataIndex + 1] << 8);
            dataIndex += 2;
        } else {
            // Single byte run: bit 0 = value, bits 1-6 = count
            value = firstByte & 0x01;
            count = firstByte >> 1;
        }

        // Apply the run
        for (let i = 0; i < count && z < depth; i++) {
            if (x < width) {
                tiles[x][z] = value;
            }
            x++;
            if (x >= storedWidth) {
                x = 0;
                z++;
            }
        }
    }

    return tiles;
}

function findFloorPositions(tiles, width, depth) {
    const positions = [];
    for (let x = 0; x < width; x++) {
        for (let z = 0; z < depth; z++) {
            if (tiles[x][z] === 1) {
                positions.push([x, z]);
            }
        }
    }
    return positions;
}

function distance(pos1, pos2) {
    return Math.sqrt(Math.pow(pos1[0] - pos2[0], 2) + Math.pow(pos1[1] - pos2[1], 2));
}

function getTierForDistance(dist) {
    if (dist < 30) return 1;
    if (dist < 60) return 2;
    if (dist < 100) return 3;
    if (dist < 150) return 4;
    return 5;
}

function isPositionClear(pos, existingPositions, minDist = 3) {
    for (const existing of existingPositions) {
        if (distance(pos, existing) < minDist) {
            return false;
        }
    }
    return true;
}

function randomChoice(arr) {
    return arr[Math.floor(Math.random() * arr.length)];
}

function shuffle(array) {
    const arr = [...array];
    for (let i = arr.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [arr[i], arr[j]] = [arr[j], arr[i]];
    }
    return arr;
}

function populateMap(inputPath, outputPath) {
    console.log(`Loading map from ${inputPath}`);

    const mapData = JSON.parse(fs.readFileSync(inputPath, 'utf8'));

    const width = mapData.width || 100;
    const depth = mapData.depth || 100;
    const spawnPos = [
        mapData.spawnPosition?.x || Math.floor(width / 2),
        mapData.spawnPosition?.z || Math.floor(depth / 2)
    ];

    console.log(`Map size: ${width}x${depth}, Spawn: (${spawnPos[0]}, ${spawnPos[1]})`);

    // Decode tiles
    const tileData = mapData.tileData || '';
    if (!tileData) {
        console.log("No tile data found!");
        return;
    }

    const tiles = decodeTileData(tileData, width, depth);
    let floorPositions = findFloorPositions(tiles, width, depth);
    console.log(`Found ${floorPositions.length} floor tiles`);

    // Track placed object positions
    const placedPositions = [[...spawnPos]];

    // Add existing enemy positions
    for (const enemy of (mapData.enemies || [])) {
        placedPositions.push([enemy.position.x, enemy.position.z]);
    }

    const newEnemies = [];
    const newProps = [];

    // Shuffle floor positions
    floorPositions = shuffle(floorPositions);

    // Place monsters - about 1 per 100 tiles
    const targetMonsters = Math.floor(floorPositions.length / 100);
    let monstersPlaced = 0;

    for (const pos of floorPositions) {
        if (monstersPlaced >= targetMonsters) break;

        const dist = distance(pos, spawnPos);

        // Skip positions too close to spawn (safe zone)
        if (dist < 15) continue;

        // Check if position is clear
        if (!isPositionClear(pos, placedPositions, 5)) continue;

        const tier = getTierForDistance(dist);

        // Select monster based on tier
        let monsterType;
        if (tier === 1) monsterType = randomChoice(TIER1_MONSTERS);
        else if (tier === 2) monsterType = randomChoice(TIER2_MONSTERS);
        else if (tier === 3) monsterType = randomChoice(TIER3_MONSTERS);
        else if (tier === 4) monsterType = randomChoice(TIER4_MONSTERS);
        else monsterType = randomChoice(TIER5_MONSTERS);

        // Occasionally spawn boss in far areas
        let isBoss = false;
        if (dist > 120 && Math.random() < 0.02) {
            monsterType = randomChoice(BOSSES);
            isBoss = true;
        }

        newEnemies.push({
            type: monsterType,
            roomId: -1,
            position: { x: pos[0], z: pos[1] },
            level: Math.max(1, tier),
            isBoss: isBoss,
            rotationY: Math.random() * 6.28
        });

        placedPositions.push(pos);
        monstersPlaced++;
    }

    console.log(`Added ${monstersPlaced} monsters`);

    // Place props - about 1 per 50 tiles
    const targetProps = Math.floor(floorPositions.length / 50);
    let propsPlaced = 0;

    for (const pos of floorPositions) {
        if (propsPlaced >= targetProps) break;

        const dist = distance(pos, spawnPos);

        // Check if position is clear
        if (!isPositionClear(pos, placedPositions, 2)) continue;

        // Select prop based on distance
        let props;
        if (dist < 30) props = [...DUNGEON_PROPS, ...CAMP_PROPS];
        else if (dist < 80) props = [...DUNGEON_PROPS, ...NATURE_PROPS];
        else if (dist < 130) props = [...DUNGEON_PROPS, ...SPOOKY_PROPS, ...TREASURE_PROPS];
        else props = [...SPOOKY_PROPS, ...TREASURE_PROPS];

        const propType = randomChoice(props);

        newProps.push({
            type: propType,
            x: pos[0] + (Math.random() - 0.5) * 0.6,
            y: 0.0,
            z: pos[1] + (Math.random() - 0.5) * 0.6,
            rotationY: Math.random() * 6.28,
            scale: 0.8 + Math.random() * 0.4
        });

        placedPositions.push(pos);
        propsPlaced++;
    }

    console.log(`Added ${propsPlaced} props`);

    // Add monster groups (clusters of related monsters)
    const groupsToAdd = Math.min(20, Math.floor(floorPositions.length / 500));
    let groupsPlaced = 0;

    for (let i = 0; i < floorPositions.length && groupsPlaced < groupsToAdd; i += 50) {
        const pos = floorPositions[i];
        const dist = distance(pos, spawnPos);

        if (dist < 40 || !isPositionClear(pos, placedPositions, 10)) continue;

        // Create a small cluster of 3-5 monsters
        const clusterSize = 3 + Math.floor(Math.random() * 3);
        const tier = getTierForDistance(dist);

        let monsterPool;
        if (tier <= 2) monsterPool = [...TIER1_MONSTERS, ...TIER2_MONSTERS];
        else if (tier <= 3) monsterPool = [...TIER2_MONSTERS, ...TIER3_MONSTERS];
        else monsterPool = [...TIER3_MONSTERS, ...TIER4_MONSTERS];

        for (let c = 0; c < clusterSize; c++) {
            const offsetX = (Math.random() - 0.5) * 8;
            const offsetZ = (Math.random() - 0.5) * 8;
            const monsterX = Math.floor(pos[0] + offsetX);
            const monsterZ = Math.floor(pos[1] + offsetZ);

            if (monsterX < 0 || monsterX >= width) continue;
            if (monsterZ < 0 || monsterZ >= depth) continue;
            if (tiles[monsterX][monsterZ] !== 1) continue;

            newEnemies.push({
                type: randomChoice(monsterPool),
                roomId: -1,
                position: { x: monsterX, z: monsterZ },
                level: Math.max(1, tier),
                isBoss: false,
                rotationY: Math.random() * 6.28
            });
        }

        placedPositions.push(pos);
        groupsPlaced++;
    }

    console.log(`Added ${groupsPlaced} monster groups`);

    // Merge with existing data
    mapData.enemies = [...(mapData.enemies || []), ...newEnemies];
    mapData.placedProps = [...(mapData.placedProps || []), ...newProps];
    mapData.name = (mapData.name || 'Map') + ' (Populated)';

    // Save
    fs.writeFileSync(outputPath, JSON.stringify(mapData, null, 2));

    console.log(`Saved populated map to ${outputPath}`);
    console.log(`Total enemies: ${mapData.enemies.length}`);
    console.log(`Total props: ${mapData.placedProps.length}`);
}

// Run
const inputMap = path.join('C:/Claude/SafeRoom3D/maps', 'Steves Place.json');
const outputMap = path.join('C:/Claude/SafeRoom3D/maps', 'Steves Place (Populated).json');

populateMap(inputMap, outputMap);

File: populate_map.py
#!/usr/bin/env python3
"""
Populate a SafeRoom3D map with aesthetically placed monsters and props.
Uses the map's tile data to intelligently place content.
"""

import json
import base64
import gzip
import random
from pathlib import Path

# Monster types by difficulty tier
TIER1_MONSTERS = ["dungeon_rat", "slime", "goblin"]
TIER2_MONSTERS = ["goblin_shaman", "goblin_thrower", "spider", "mushroom", "bat"]
TIER3_MONSTERS = ["skeleton", "wolf", "lizard", "eye", "badlama"]
TIER4_MONSTERS = ["crawler_killer", "shadow_stalker", "mimic", "flesh_golem"]
TIER5_MONSTERS = ["living_armor", "plague_bearer", "lava_elemental", "void_spawn"]

BOSSES = ["skeleton_lord", "dragon_king", "spider_queen", "the_butcher", "mordecai", "mongo"]

# Prop types by theme
DUNGEON_PROPS = ["barrel", "crate", "pot", "torch", "bone_pile", "skull_pile", "rubble_heap"]
TREASURE_PROPS = ["chest", "treasure_chest", "scattered_coins", "ancient_scroll"]
SPOOKY_PROPS = ["blood_pool", "coiled_chains", "manacles", "discarded_sword", "forgotten_shield"]
NATURE_PROPS = ["moss_patch", "glowing_mushrooms", "water_puddle", "thorny_vines"]
CAMP_PROPS = ["campfire", "abandoned_campfire", "rat_nest", "moldy_bread"]

def decode_tile_data(tile_data: str, width: int, depth: int) -> list:
    """Decode RLE+Base64 tile data to 2D array."""
    # Decode base64
    compressed = base64.b64decode(tile_data)
    # Decompress gzip
    decompressed = gzip.decompress(compressed)

    # RLE decode
    tiles = [[0] * depth for _ in range(width)]
    x, z = 0, 0
    i = 0
    while i < len(decompressed) and x < width:
        value = decompressed[i]
        i += 1
        count = 1
        if i < len(decompressed) and decompressed[i] == 0xFF:
            i += 1
            if i < len(decompressed):
                count = decompressed[i]
                i += 1

        for _ in range(count):
            if x < width and z < depth:
                tiles[x][z] = value
                z += 1
                if z >= depth:
                    z = 0
                    x += 1

    return tiles

def find_floor_positions(tiles: list, width: int, depth: int) -> list:
    """Find all floor positions (value 1)."""
    positions = []
    for x in range(width):
        for z in range(depth):
            if tiles[x][z] == 1:
                positions.append((x, z))
    return positions

def distance_from_spawn(pos, spawn_pos):
    """Calculate distance from spawn point."""
    return ((pos[0] - spawn_pos[0])**2 + (pos[1] - spawn_pos[1])**2) ** 0.5

def get_tier_for_distance(dist: float) -> int:
    """Get monster tier based on distance from spawn."""
    if dist < 30: return 1
    elif dist < 60: return 2
    elif dist < 100: return 3
    elif dist < 150: return 4
    else: return 5

def is_position_clear(pos, existing_positions, min_dist=3):
    """Check if position is clear of other objects."""
    for existing in existing_positions:
        if distance_from_spawn(pos, existing) < min_dist:
            return False
    return True

def populate_map(input_path: str, output_path: str):
    """Main function to populate a map."""
    print(f"Loading map from {input_path}")

    with open(input_path, 'r') as f:
        map_data = json.load(f)

    width = map_data.get('width', 100)
    depth = map_data.get('depth', 100)
    spawn_pos = (
        map_data.get('spawnPosition', {}).get('x', width // 2),
        map_data.get('spawnPosition', {}).get('z', depth // 2)
    )

    print(f"Map size: {width}x{depth}, Spawn: {spawn_pos}")

    # Decode tiles
    tile_data = map_data.get('tileData', '')
    if not tile_data:
        print("No tile data found!")
        return

    tiles = decode_tile_data(tile_data, width, depth)
    floor_positions = find_floor_positions(tiles, width, depth)
    print(f"Found {len(floor_positions)} floor tiles")

    # Track placed object positions
    placed_positions = list(spawn_pos)  # Start with spawn

    # Add existing enemy positions
    for enemy in map_data.get('enemies', []):
        placed_positions.append((enemy['position']['x'], enemy['position']['z']))

    new_enemies = []
    new_props = []

    # Randomly sample floor positions for monster/prop placement
    random.shuffle(floor_positions)

    # Place monsters - about 1 per 100 tiles
    target_monsters = len(floor_positions) // 100
    monsters_placed = 0

    for pos in floor_positions:
        if monsters_placed >= target_monsters:
            break

        dist = distance_from_spawn(pos, spawn_pos)

        # Skip positions too close to spawn (safe zone)
        if dist < 15:
            continue

        # Check if position is clear
        if not is_position_clear(pos, placed_positions, min_dist=5):
            continue

        tier = get_tier_for_distance(dist)

        # Select monster based on tier
        if tier == 1:
            monster_type = random.choice(TIER1_MONSTERS)
        elif tier == 2:
            monster_type = random.choice(TIER2_MONSTERS)
        elif tier == 3:
            monster_type = random.choice(TIER3_MONSTERS)
        elif tier == 4:
            monster_type = random.choice(TIER4_MONSTERS)
        else:
            monster_type = random.choice(TIER5_MONSTERS)

        # Occasionally spawn boss in far areas
        is_boss = False
        if dist > 120 and random.random() < 0.02:
            monster_type = random.choice(BOSSES)
            is_boss = True

        new_enemies.append({
            "type": monster_type,
            "roomId": -1,
            "position": {"x": pos[0], "z": pos[1]},
            "level": max(1, tier),
            "isBoss": is_boss,
            "rotationY": random.uniform(0, 6.28)
        })

        placed_positions.append(pos)
        monsters_placed += 1

    print(f"Added {monsters_placed} monsters")

    # Place props - about 1 per 50 tiles
    target_props = len(floor_positions) // 50
    props_placed = 0

    for pos in floor_positions:
        if props_placed >= target_props:
            break

        dist = distance_from_spawn(pos, spawn_pos)

        # Check if position is clear
        if not is_position_clear(pos, placed_positions, min_dist=2):
            continue

        # Select prop based on distance (more treasure/spooky far from spawn)
        if dist < 30:
            props = DUNGEON_PROPS + CAMP_PROPS
        elif dist < 80:
            props = DUNGEON_PROPS + NATURE_PROPS
        elif dist < 130:
            props = DUNGEON_PROPS + SPOOKY_PROPS + TREASURE_PROPS
        else:
            props = SPOOKY_PROPS + TREASURE_PROPS

        prop_type = random.choice(props)

        new_props.append({
            "type": prop_type,
            "x": float(pos[0]) + random.uniform(-0.3, 0.3),
            "y": 0.0,
            "z": float(pos[1]) + random.uniform(-0.3, 0.3),
            "rotationY": random.uniform(0, 6.28),
            "scale": random.uniform(0.8, 1.2)
        })

        placed_positions.append(pos)
        props_placed += 1

    print(f"Added {props_placed} props")

    # Add monster groups (clusters of related monsters)
    groups_to_add = min(20, len(floor_positions) // 500)
    groups_placed = 0

    for pos in floor_positions[::50]:  # Sample every 50th position
        if groups_placed >= groups_to_add:
            break

        dist = distance_from_spawn(pos, spawn_pos)
        if dist < 40 or not is_position_clear(pos, placed_positions, min_dist=10):
            continue

        # Create a small cluster of 3-5 monsters
        cluster_size = random.randint(3, 5)
        tier = get_tier_for_distance(dist)

        if tier <= 2:
            monster_pool = TIER1_MONSTERS + TIER2_MONSTERS
        elif tier <= 3:
            monster_pool = TIER2_MONSTERS + TIER3_MONSTERS
        else:
            monster_pool = TIER3_MONSTERS + TIER4_MONSTERS

        for i in range(cluster_size):
            offset_x = random.uniform(-4, 4)
            offset_z = random.uniform(-4, 4)
            monster_pos = (int(pos[0] + offset_x), int(pos[1] + offset_z))

            if monster_pos[0] < 0 or monster_pos[0] >= width:
                continue
            if monster_pos[1] < 0 or monster_pos[1] >= depth:
                continue
            if tiles[monster_pos[0]][monster_pos[1]] != 1:
                continue

            new_enemies.append({
                "type": random.choice(monster_pool),
                "roomId": -1,
                "position": {"x": monster_pos[0], "z": monster_pos[1]},
                "level": max(1, tier),
                "isBoss": False,
                "rotationY": random.uniform(0, 6.28)
            })

        placed_positions.append(pos)
        groups_placed += 1

    print(f"Added {groups_placed} monster groups")

    # Merge with existing data
    map_data['enemies'] = map_data.get('enemies', []) + new_enemies
    map_data['placedProps'] = map_data.get('placedProps', []) + new_props
    map_data['name'] = map_data.get('name', 'Map') + ' (Populated)'

    # Save
    with open(output_path, 'w') as f:
        json.dump(map_data, f, indent=2)

    print(f"Saved populated map to {output_path}")
    print(f"Total enemies: {len(map_data['enemies'])}")
    print(f"Total props: {len(map_data['placedProps'])}")

if __name__ == '__main__':
    input_map = Path('C:/Claude/SafeRoom3D/maps/Steves Place.json')
    output_map = Path('C:/Claude/SafeRoom3D/maps/Steves Place (Populated).json')

    populate_map(str(input_map), str(output_map))
